spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/sampler/PercentageBasedSamplerTests.java

```diff
@@ -49,19 +49,36 @@ public void should_pass_given_percent_of_samples() throws Exception {
 				withPercentage(3));
 	}
 
+	@Test
+	public void should_pass_given_percent_of_samples_with_fractional_element() throws Exception {
+		int numberOfIterations = 100000;
+		float percentage = 0.35f;
+		this.samplerConfiguration.setPercentage(percentage);
+
+		int numberOfSampledElements = countNumberOfSampledElements(numberOfIterations);
+
+		int threshold = (int) (numberOfIterations * percentage);
+		then(numberOfSampledElements).isCloseTo(threshold,
+				withPercentage(3));
+	}
+
 	private int countNumberOfSampledElements(int numberOfIterations) {
 		int passedCounter = 0;
 		for (int i = 0; i < numberOfIterations; i++) {
 			boolean passed = new PercentageBasedSampler(this.samplerConfiguration)
-					.isSampled(this.span);
+					.isSampled(newSpan());
 			passedCounter = passedCounter + (passed ? 1 : 0);
 		}
 		return passedCounter;
 	}
 
 	@Before
 	public void setupSpan() {
-		this.span = Span.builder().traceId(RANDOM.nextLong()).build();
+		this.span = newSpan();
+	}
+
+	Span newSpan() {
+		return Span.builder().traceId(RANDOM.nextLong()).build();
```

Comment:
yep. most importantly the sampler requires trace ids to be random, and.. the sample rate isn't stable until about 100,000 random ids pass through.

This means we might want to consider re-introducing a counting sampler, or making an option to make this one a counting one. If you have a low-traffic server, counting could be appropriate, given the warts.

@eirslett @kristofa I was thinking about this for brave, too.. wdyt?